### PR TITLE
BXMSDOC-2012 (for in-fork master-final): Compile and update new decision tables doc for DM 7.0 GA

### PR DESCRIPTION
**Pull request by sterobin**
Links:
- [JIRA](https://issues.jboss.org/browse/BXMSDOC-2012) (see for details)
- [Doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-2012/)

### CODE REVIEW

File: docs/product-user-guide/src/main/asciidoc/decision-tables-attributes-ref.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-attributes-ref.adoc]: @sterobin If the user uploads multiple spreadsheets into the same package and more than one has a RuleTable that uses the same attributes listed in "Table 2.3" then there is definitely room for error. These attributes, when added to the DRL, control Drools rule engine behaviour for the whole package. If XLS1 defines, for example, ENABLED as false and XLS2 defines it as DISABLED then which takes precedence will rely on the order in which the different XLSs are added to the rule engine. So a warning in the documentation should be given about using the attributes in "Table 2.3" in different RuleTables in different XLSs in the same package. If the User uses the same attributes in different XLSs it would be advisable to have them in different packages. The regular RuleTable attributes ("Table 2.1") only apply to the XLS being processed and not the whole package. To be honest I've never seen Table 2.1's attributes used in the RuleTable and had to run some tests here to verify they actually worked! (they do).

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: @sterobin You are pushing my knowledge! I however checked some Unit Tests we have and found an XLS where a CONDITION column does not have a OBJECT TYPE cell populated but has a full expression, including an object type, in the CONSTRAINT cell (e.g. the constraint value was ```Person( age == $1 )```). If a CONDITION column has ALL the other rows (OBJECT_TYPE and CONSTRAINT) empty but the the column has data for interpolation then compiling of the XLS will fail. So although the documents state OBJECT_TYPE and CONSTRAINT are optional in reality you need to provide one or the other or both.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: @sterobin I cannot explain the history of why we support custom codes beginning with the initial character and I've never seen it used to be honest.. however I guess documentation should reflect what is possible and not question whether the feature makes sense. This predates my arrival at Red Hat.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: @sterobin I'd be inclined to remove (re-word) https://github.com/sterobin/kie-docs/pull/10/files#diff-82b1ecd3589462e9deb4a8011f6f3ba4R21 and https://github.com/sterobin/kie-docs/pull/10/files#diff-82b1ecd3589462e9deb4a8011f6f3ba4R22 as what is written there only applies to CONDITION columns and you state the same, but in more detail, here. Perhaps the first occurrences I link need to state the use of the two rows in between "Rule attributes" and "Column label" depend upon the type of Rule (column) Attribute?

Comment (manstis) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: @sterobin My comments above (your "(1 of 6)") should be integrated into the text here. I think that'd clarify it.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-set-entries-ref.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-set-entries-ref.adoc]: @sterobin I reviewed the code and, as odd as it may seem, is indeed true.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: @sterobin It means RuleSet. The reason being is that the "package" name comes from RuleSet and then "rule" names in the resultant DRL come from the RuleTable; so I could have ```package A rule Rules_1``` from one XLS (where the RuleSet is "A" and the RuleTable is "Rules") and ```package A rule Rule_1``` from another XLS **IF** it too had a RuleSet of "A" and RuleTable of "Rules"). I believe this would cause a compilation error. You can have the same rule names in different packages or different rule names in the same package. I checked the code and neither spreadsheet file name or worksheet name are used.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: I guess I disagree with you a bit. :) Some of the greatest product improvements, bug fixes, etc., spawn from the documentation process. Documenting a product reveals is weaknesses and ambiguities, before they get to the user. I'll think more on it and see if it's better to leave it or remove it. I was just curious how used it is. Thanks!

Comment (manstis) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: @sterobin It was not my intention to cause offence. If I have, I apologise.

Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: Oh geeze, now I feel bad that I came off that way. Not at all! Sorry if I made it sound that way.

**Comment by sterobin** [comment]
@manstis, I've implemented your answers to my initial questions and have added comments for you in the doc, to approve the resulting updates or to answer a follow-up question. You can search the [doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-2012/) for `@Michael`. There are 5 comments.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: @sterobin A single workbook per package (RuleSet) is encouraged (for most Users) but not essential (if you know what you're doing). Therefore IDK whether "IMPORTANT: You should create only one...." is too strong wording vs "IMPORTANT: You are encouraged to create only one...". Final decision with you!

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: @sterobin Not entirely true (and I think my last comment was probably misleading); if you have a value in the "pattern" cell (first row under CONDITION) it cannot contain a "code snippet" (e.g. ```Person( age == $param )```) as it gives a compilation error. You'd need ```Person``` in the "pattern" cell and ```age == $param``` (or simply ```age```) in the "constraint" cell (second row under CONDITION). However you can leave the "pattern" cell empty any put ```Person( age == $param )``` in ONLY the "constraint" cell. 

So, I think in summary the original over-use of "optional" should be:

- Pattern cell: Optional (IF Constraint cell is a full expression) otherwise optional.
- Constraint cell: Compulsory.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: @sterobin Yes, fine now.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-set-entries-ref.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-set-entries-ref.adoc]: @sterobin No, only the attributes that can be used in RuleTable have "Initial" support (weird as that is too!)

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (manstis) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: @sterobin Looks good to me. Thanks.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: Got it. Will consider slight rewording.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: Aaah, I see. K. Makes more sense. Will adjust. Thanks.

**Comment by sterobin** [comment]
@manstis, I've implemented your last round of comments and have updated the [doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-2012/) accordingly. Do you have any other feedback for this document/PR or do you approve of it now as is?

**Review by manstis — APPROVED** [review]
Thanks @sterobin

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: "The name string is used".. Probably fine, but may be better to say "The name (type: string) is used"..

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: Should "Save" be bolded? Not sure as this is the excel spreadsheet UI.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-attributes-ref.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-attributes-ref.adoc:17]: Probably not an issue, but just to be sure. Does the product actually use the term "salience"? My only concern would be how translatable it would be.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-set-entries-ref.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-set-entries-ref.adoc]: I would remove the `"` after the work mark. The example makes is clear, but the standalone after the work mark looks strange. Your call of course.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: "interpolating" --- REALLY? Who knows this word?  :)

Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: Style guide says no go for "etc.,". Can you replace with "and so on"?

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: Style guide says no go for "etc.,". Can you replace with "and so on"?

Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: What did I tell you about "interpolating"? Who knows this word? :)

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: "successive" is another that I pause on for translation potential issues. Would "consecutive" be better?

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (michelehaglund) [inline docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: "successive" is another that I pause on for translation potential issues. Would "consecutive" be better?

File: docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-defining-proc.adoc]: No. Only bolded when it references the UI, which this doesn't. Just saying generally to save.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-attributes-ref.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-attributes-ref.adoc:17]: Yes. It's one of the available rule attributes.

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: Ha! It's one of those standard words when dealing with decision table logic (decision tables are a common thing in general for coding decision logic). And I guess I just got indoctrinated by all the upstream documentation. I'll think on it. Thanks.

Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: You're correct. Thanks!

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (sterobin) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: Sure.

**Comment by sterobin** [comment]
Thanks @michelehaglund! Since a lot of the work on this was based on upstream documentation, I missed a lot of little things while working out much bigger ones. Thank you so much for your keen eye!

File: docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc
Comment (manstis) [reply docs/product-user-guide/src/main/asciidoc/decision-tables-rule-table-entries-ref.adoc]: @sterobin @michelehaglund Wait until you write/look at the documents relating to decision table validation and verification... this uses the word "subsumption" -- and community documents may even mention "subsumptive" (there's a "made up" word if ever I read one!) ;-)

**Pull request merged**
